### PR TITLE
[SYCL] Add generation of device image with specialization constants replaced by default values

### PR DESCRIPTION
**Pull request by maksimsab**

### CODE REVIEW

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 1]
```diff
@@ -542,6 +548,15 @@ std::string saveModuleProperties(module_split::ModuleDesc &MD,
     PropSet.add(PropSetRegTy::SYCL_HOST_PIPES, HostPipePropertyMap);
   }
 
+  if (MD.isSpecConstantDefault()) {
+    PropSet[PropSetRegTy::SYCL_MISC_PROP].insert({"defaultSpecConstants", 1});
+    const std::optional<std::string> &OriginalImageName =
+        MD.tryGetOriginalImageName();
+    assert(OriginalImageName.has_value() && "Original image name must be set");
+    PropSet[PropSetRegTy::SYCL_MISC_PROP].insert(
+        {"originalImage", StringRef(*OriginalImageName)});
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:557]: Why do we need that? Can't runtime just select any other device image, which is compatible with the device and has necessary kernels?

Please note that device image names do not leave `sycl-post-link`, they are not available to runtime and therefore this property is useless there.

Having the first property you add should be enough for RT. We could even rename it to something like `nonModifiableSpecConstants`

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+; CHECK-PROP0-NOT: originalImage
+
+; CHECK-PROP1: defaultSpecConstants=1|1
+
+; CHECK-IR0: call i32 @_Z20__spirv_SpecConstantii
+; CHECK-IR0: call %struct.B @_Z29__spirv_SpecConstantCompositeiii_Rstruct.B
+; CHECK-IR0: call %struct.A @_Z29__spirv_SpecConstantCompositeistruct.B_Rstruct.A
+
+; CHECK-IR1-NOT: SpecConstant
+; CHECK-IR1: call void @llvm.memcpy.p4i8.p4i8.i64(i8 addrspace(4)* align 8 %1, i8 addrspace(4)* align 8 addrspacecast (i8 addrspace(1)* bitcast (%"class.sycl::_V1::specialization_id" addrspace(1)* @_ZL1c to i8 addrspace(1)*) to i8 addrspace(4)*), i64 16, i1 false)
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll:23]: Have you performed experiments to ensure that this `memcpy` can successfully be optimized out by some constant propagation passes? Because if not, then we can't emit it here.

Moreover, I would suggest not to emit it right away, but instead to plain `store` of a constant literal value, which we should be able to obtain from the global variable

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 3]
```diff
@@ -629,6 +651,16 @@ void EntryPointGroup::rebuildFromNames(const std::vector<std::string> &Names,
   });
 }
 
+void EntryPointGroup::rebuild(const Module &M) {
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:654]: Where is this method used?

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 4]
```diff
@@ -972,6 +1016,11 @@ processInputModule(std::unique_ptr<Module> M) {
     DUMP_ENTRY_POINTS(MDesc.entries(), MDesc.Name.c_str(), 1);
 
     MDesc.fixupLinkageOfDirectInvokeSimdTargets();
+    SmallVector<module_split::ModuleDesc, 2> MMsWithDefaultSpecConsts;
+    if (GenerateDeviceImageWithDefaultSpecConsts)
+      MMsWithDefaultSpecConsts = generateModuleDescWithDefaultSpecConstants(
+          MDesc, Modified, SplitOccurred);
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:1022]: Emitting that new special device image should be cheaper than handling ESIMD. Therefore, this step should happen _after_ ESIMD handling to avoid doing ESIMD handling for an extra module once again.

Moreover, we likely have a bug in `handleESIMD`:

https://github.com/intel/llvm/blob/c4c9bdbd4efc24b3259025fde8ec9e9d0611f7ed/llvm/tools/sycl-post-link/sycl-post-link.cpp#L875-L880

Spec constants are handled _before_ we may do a merge of two device images. If both ESIMD and non-ESIMD device images contain spec constants, then merging them _after_ handling spec constants would produce an invalid module. Metadata about default values would be incorrectly merged, different spec constants would have assigned the same IDs, etc.

I would suggest the following code structure:

```c++
while (Splitter->hasMoreSplits()) {
  do_split();
  MMs = handleESIMD();
  for (int I = 0, E = MMs.size(); I < E; ++I) { // we only want to process original content of MMs
    processSpecConstants(MMs, I);
  }

  for (M : MMs) {
    saveModule(M)
  }
}

processSpecConstants(MMs, Ind) {
  if (!LowerSpecConstants)
    return;
  auto &M = MMs[Ind];
  if (!M->hasSpecConstants())
    return;
  if (GenerateExtraImage) {
    M2 = Clone(M);
    SpecConstantsPass(M2, mode=hardcode_defaults);
    M2.SpecialFlag = true;
    MMs.push_back(M2);
  }
  SpecConstantsPass(M, mode=mode_from_arguments);
}
```

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 5]
```diff
@@ -914,3 +919,64 @@ bool SpecConstantsPass::collectSpecConstantDefaultValuesMetadata(
...
+
+/// Function generates a copy of Module and replaces all spec constant uses
+/// memcpy intrinsic.
+/// If the module doesn't contain spec constants then nullptr is returned.
+std::unique_ptr<Module>
+llvm::generateDeviceImageWithDefaultSpecConstants(const Module &M) {
+  if (!checkModuleContainsSpecConsts(M))
+    return nullptr;
+
+  std::unique_ptr<Module> M2 = CloneModule(M);
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:960]: Cloning modules should not be a part of a pass. The pass handles spec constants in a provided module. If we need to create several modules, that should be done outside of the pass scope.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 6]
```diff
@@ -914,3 +919,64 @@ bool SpecConstantsPass::collectSpecConstantDefaultValuesMetadata(
...
+        !F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL))
+      continue;
+
+    SpecConstDeclarations.push_back(&F);
+    for (User *U : F.users())
+      if (auto *CI = dyn_cast<CallInst>(U))
+        CIs.push_back(CI);
+  }
+
+  replaceSpecConstsWithDefaultValue(*M2, CIs);
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:977]: This kind of replicates existing pass logic. I would suggest that we replace "bool SetValAtRT = true" with some `enum`, which would have 3 possible values:
- `native` - emit SPIR-V intrinsics
- `emulation` - use kernel arguments
- `default` - replace with default values

Then, we will simply be able to run the pass in different modes. It should be updated to have a 3rd main branch, which we kind of had for SYCL 1.2.1 (in AOT mode, they values were replaced with defaults)

**Review by AlexeySachkov — CHANGES_REQUESTED** [review]

**Comment by maksimsab** [comment]
TODO:
- [ ] add test with padding

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 5 above)
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:960]: Yes, I understand, but I still think that it would be better to have cloning logic outside of `SpecConstants.cpp`, because strictly speaking, it is not related to `__syclGetSpecializationConstant*` intrinsics handling

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll (hunk 2 above)
Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll]: Moved to `store` instruction instead of `memcpy`.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp (hunk 3 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: It is a leftover from the previous version. Removed it.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 7]
```diff
@@ -562,6 +562,21 @@ void ModuleDesc::cleanup() {
...
+void ModuleDesc::setSpecConstantDefault(bool Value) {
+  Props.IsSpecConstantDefault = Value;
+}
+
+ModuleDesc CloneModuleDesc(const ModuleDesc &MD) {
+  std::unique_ptr<Module> NewModule = CloneModule(MD.getModule());
+  ModuleDesc NewMD(std::move(NewModule));
+  NewMD.EntryPoints.Props = MD.EntryPoints.Props;
+  return NewMD;
+}
```
Comment (againull) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:578]: It seems like some kind of deep copy of ModuleDesc is needed here. At least in addition to EntryPoints.Props also EntryPoints.Functions need to be copied.
Otherwise cloned module/device image is not registered in ProgramManager through void ProgramManager::addImages(pi_device_binaries DeviceBinary)  because it doesn't have entry points.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 8]
```diff
@@ -621,12 +730,16 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
   MapVector<StringRef, MDNode *> SCMetadata;
   SmallVector<MDNode *, 4> DefaultsMetadata;
 
+  bool IRModified = false;
+  if (Mode == HandlingMode::default_values) {
+    IRModified |= replaceSpecConstsWithDefaultValues(M);
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:735]: Why can't we integrate this into a regular flow of the pass? The PR duplicates some of a basic functionality, increasing maintenance cost. I though that we could simply add a new branch next to `if (Mode == HandlingMode::native)`

New path does not generate metadata with default spec constant values - I would double-check if it is indeed not used by runtime. Could be the case that it is used to handle `get_specialization_constant_value` on host.

File: llvm/tools/sycl-post-link/ModuleSplitter.h [hunk 9]
```diff
@@ -183,12 +184,19 @@ class ModuleDesc {
   // Cleans up module IR - removes dead globals, debug info etc.
   void cleanup();
 
+  bool isSpecConstantDefault() const;
+  void setSpecConstantDefault(bool Value);
+
+  friend ModuleDesc CloneModuleDesc(const ModuleDesc &MD);
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/ModuleSplitter.h:190]: Why not make this into a method `clone`?

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/struct-with-padding.ll [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+; RUN: cat %t.table | FileCheck %s -check-prefix=CHECK-TABLE -DPATH=%t
+; RUN: cat %t_1.prop | FileCheck %s -check-prefix=CHECK-PROP1
+; RUN: cat %t_1.ll | FileCheck %s -check-prefix=CHECK-IR1
+
+; CHECK-TABLE: [[PATH]]_0.ll|[[PATH]]_0.prop|[[PATH]]_0.sym
+; CHECK-TABLE: [[PATH]]_1.ll|[[PATH]]_1.prop|[[PATH]]_1.sym
+
+; CHECK-PROP1: specConstsReplacedWithDefault=1|1
+
+; CHECK-IR1-NOT: SpecConstant
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/default-value/struct-with-padding.ll:13]: Better to use `--implicit-check-not`, because `NOT` directives are scoped, i.e. this line would allow `SpecConstant` to appear after the `store` you are looking for

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 8 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:735]: At the first glance, regular flow of the pass is hard to read and I didn't want to make it worse.

Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:735]: Well, it might be easy to make it good.

Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:735]: Yeah, I agree that that function is way too long, but perhaps it is time to outline branches of that main `if` into a separate functions to make high-level logic easier to read

Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:735]: Started it here: https://github.com/intel/llvm/pull/10613

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 11]
```diff
@@ -612,6 +617,110 @@ Instruction *emitSpecConstantRecursive(Type *Ty, Instruction *InsertBefore,
...
+/// Example:
+///  %"spec_id" = type { %A }
+///  %A = type { i32 }
+///  @value = constant %"spec_id" { %A { i32 1 } }, align 4
+///
+///  define spir_func void @func() {
+///  entry:
+///    %a.i = alloca %A, align 4
+///    %a.ascast.i = addrspacecast %A* %a.i to %A addrspace(4)*
+///    call spir_func void @getCompositeSpecConst(%a.ascast.i, %2, @value, %3)
```
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:662]: Added such case in the test.

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 8 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp]: Integrated it into the regular loop.

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp (hunk 7 above)
Comment (maksimsab) [reply llvm/tools/sycl-post-link/ModuleSplitter.cpp]: Addresses it. Now symbol generation is being checked in `llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll`

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 12]
```diff
@@ -209,6 +209,12 @@ cl::opt<bool> DeviceGlobals{
     cl::desc("Lower and generate information about device global variables"),
     cl::cat(PostLinkCat)};
 
+cl::opt<bool> GenerateDeviceImageWithDefaultSpecConsts{
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:212]: I've just realized that this option is likely incompatible with `--ir-output-only`, let's add a corresponding check to `main`.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 13]
```diff
@@ -719,6 +724,26 @@ void updatePaddingInLastMDNode(LLVMContext &Ctx,
...
+/// the given Value @Dst.
+/// Note: Types of values Dst and V might differ because of padding bytes
+/// inserted by Clang FE.
+/// For example:
+/// Type of specialization constant might be <{ i32, i8, [ 3 x i8 ] }>, where
+/// the last component are padding bytes.
+/// specialization id in this case could be { i32, i8 } { i32 1, i8 1 }.
+/// As you can see, padding bytes are absent. In order to mitigate this we
+/// perform bitcast from specialization id type to specialization constant
+/// type.
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:737]: I wonder how will it work if a padding is inserted in the middle of a struct. **Not** asking to address in this PR, because it is known that we can't handle such structs anyway and therefore it should be addressed as a whole in a separate PR. I have initial set of changes for that [in my fork](https://github.com/AlexeySachkov/llvm/commits/private/asachkov/spec-constants-with-padding)

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 14]
```diff
@@ -735,8 +760,8 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
   // intrinsic to find its calls and lower them depending on the SetValAtRT
   // setting (see below).
   bool IRModified = false;
-
   LLVMContext &Ctx = M.getContext();
+  std::vector<Function *> SpecConstDeclarations;
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:764]: This should most likely be a `SmallVector` instead. See [Programmer's Manual](https://llvm.org/docs/ProgrammersManual.html#sequential-containers-std-vector-std-list-etc):

> Pick the first in this section that will do what you want.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 15]
```diff
@@ -14,10 +14,15 @@
...
 #include "llvm/ADT/StringMap.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/InstIterator.h"
 #include "llvm/IR/Instruction.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/Operator.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+#include <memory>
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:24]:
```suggestion
```

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/split-by-source.ll [hunk 16]
```diff
@@ -0,0 +1,54 @@
+; Test checks generation of device images for splitted kernels by source.
+
+; RUN: sycl-post-link -split=source -symbols -o %t.table %s -generate-device-image-default-spec-consts && \
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/default-value/split-by-source.ll:3]:
```suggestion
; RUN: sycl-post-link -split=source -o %t.table %s -generate-device-image-default-spec-consts
```

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/esimd.ll [hunk 17]
```diff
@@ -0,0 +1,55 @@
+; Test checks generation of device image of esimd kernel.
+
+; RUN: sycl-post-link -split=auto -emit-param-info -symbols -emit-exported-symbols -split-esimd -lower-esimd -O2 -spec-const=rt -device-globals -o %t.table %s -generate-device-image-default-spec-consts
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/default-value/esimd.ll:3]:
```suggestion
; RUN: sycl-post-link -split=auto -split-esimd -lower-esimd -O2 -spec-const=rt -o %t.table %s -generate-device-image-default-spec-consts
```

I don't see why we need all those extra options

File: llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+; RUN: sycl-post-link -split=auto -spec-const=rt -symbols -S -o %t.table %s -generate-device-image-default-spec-consts
+; RUN: FileCheck %s -input-file %t.table -check-prefix=CHECK-TABLE -DPATH=%t
+; RUN: FileCheck %s -input-file %t_0.prop -check-prefix=CHECK-PROP0
+; RUN: FileCheck %s -input-file %t_1.prop -check-prefix=CHECK-PROP1
+; RUN: FileCheck %s -input-file %t_0.ll -check-prefix=CHECK-IR0
+; RUN: FileCheck %s -input-file %t_1.ll -check-prefix=CHECK-IR1 --implicit-check-not "SpecConstant"
+; RUN: FileCheck %s -input-file %t_0.sym -check-prefix=CHECK-SYM0
+; RUN: FileCheck %s -input-file %t_1.sym -check-prefix=CHECK-SYM1
+
+; CHECK-TABLE: [[PATH]]_0.ll|[[PATH]]_0.prop|[[PATH]]_0.sym
```
Comment (AlexeySachkov) [inline llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll:14]: Note: just using `{{.*}}_0.ll` should be enough here. I guess that checking the exact path is not the main purpose of the check, but checking how many lines there are in the table is.

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 19]
```diff
@@ -878,6 +902,10 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
     }
   }
 
+  if (Mode == HandlingMode::default_values)
+    for (Function *F : SpecConstDeclarations)
+      F->removeFromParent();
```
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:907]: We do, regardless of optimization level:

https://github.com/intel/llvm/blob/7697e7bf9ab528a3efe16dbc2b168db825a96b5b/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L306-L319

https://github.com/intel/llvm/blob/7697e7bf9ab528a3efe16dbc2b168db825a96b5b/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L554-L563

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 20]
```diff
@@ -696,7 +706,9 @@ bool processSpecConstants(module_split::ModuleDesc &MD) {
   ModulePassManager RunSpecConst;
   ModuleAnalysisManager MAM;
   bool SetSpecConstAtRT = (SpecConstLower == SC_USE_RT_VAL);
-  SpecConstantsPass SCP(SetSpecConstAtRT);
+  SpecConstantsPass SCP(SetSpecConstAtRT
+                            ? SpecConstantsPass::HandlingMode::native
+                            : SpecConstantsPass::HandlingMode::emulation);
```
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Yes, I would like to outline it in separate PR.

[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Here: https://github.com/intel/llvm/pull/10741

File: llvm/tools/sycl-post-link/sycl-post-link.cpp [hunk 21]
```diff
@@ -209,6 +209,12 @@ cl::opt<bool> DeviceGlobals{
     cl::desc("Lower and generate information about device global variables"),
     cl::cat(PostLinkCat)};
 
+cl::opt<bool> GenerateDeviceImageWithDefaultSpecConsts{
+    "generate-device-image-default-spec-consts",
+    cl::desc("Generate device image which contains specialization constants "
```
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Would it be OK?
https://github.com/intel/llvm/pull/10115/commits/7af47491b7cbe57e0fed5042e6893757662aaca7#diff-7d1ae2c6e3e847852b55edea2a54fc5309328f32e56800a636a4e3d9c906948bR214

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:904]: Well, yes. However, we copy device image after this cleanup.

[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:904]: If I see it correctly the current SpecConstPass leaves abandoned function declarations which are not removed later.

[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:911]: I still don't think that removal of unused declarations should be a part of `SpecConstantsPass`, because _any_ transformation pass could leave those and having cleanup code in _each_ of them does not seem like a good design. We have a dedicated pass for cleanup, no need to duplicate its code to every pass.

You can just add `StripDeadPrototypesPass` to the pipeline when launching `SpecConstantsPass` to emit that extra device image with default values of spec constants

**Review by AlexeySachkov — APPROVED** [review]

File: llvm/tools/sycl-post-link/SpecConstants.cpp [hunk 22]
```diff
@@ -14,11 +14,14 @@
 #include "llvm/ADT/APInt.h"
 #include "llvm/ADT/StringMap.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/InstIterator.h"
 #include "llvm/IR/Instruction.h"
 #include "llvm/IR/Instructions.h"
 #include "llvm/IR/Operator.h"
 
+#include <vector>
```
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:23]: I saw there 2 usages of std::vector.

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (maksimsab) [reply llvm/tools/sycl-post-link/SpecConstants.cpp]: Added StripDeadPrototypesPass

File: llvm/tools/sycl-post-link/SpecConstants.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (AlexeySachkov) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:23]: Got it, I just though that we don't use it at all
